src/main/java/com/clevercloud/biscuitpulsar/AuthorizationProviderBiscuit.java

```diff
@@ -51,151 +55,84 @@ public AuthorizationProviderBiscuit(ServiceConfiguration conf, PulsarResources p
-        if (res.isLeft()) {
-            return CompletableFuture.failedFuture(new Exception(res.getLeft().toString()));
-        }
-
-        Verifier verifier = res.get();
-        verifier.set_time();
-        verifier.add_fact(topicFact(topicName)).get();
-        verifier.add_fact(topicOperationFact(topicName, TopicOperation.LOOKUP)).get();
-        verifier.add_rule("right(#authority, $tenant, $namespace, $topic, #lookup) <- right(#authority, #admin), topic_operation(#ambient, $tenant, $namespace, $topic, #lookup)").get();
-        verifier.add_check("check if right(#authority," + topicFactFragment(topicName) + "," + topicOperationSymbol(TopicOperation.LOOKUP)).get();
-        verifier.allow();
-
-        Either verifierResult = verifier.verify(runLimits);
-        log.debug(verifier.print_world());
-
-        if (verifierResult.isLeft()) {
-            log.debug("Biscuit canLookupAsync on [{}] NOT authorized for role [{}]: {}", topicName, role, verifierResult.getLeft());
-            return isSuperUser(role, authenticationData, this.conf);
-        } else {
-            log.debug("Biscuit canLookupAsync on [{}] authorized.", topicName);
-            return CompletableFuture.completedFuture(true);
-        }
+        return allowTopicOperationAsync(topicName, role, TopicOperation.LOOKUP, authenticationData);
     }
 
     @Override
-    //FIXME: should be reworked
     public CompletableFuture<Boolean> allowFunctionOpsAsync(NamespaceName namespaceName, String role, AuthenticationDataSource authenticationData) {
         if (!role.startsWith("biscuit:")) {
             return defaultProvider.allowFunctionOpsAsync(namespaceName, role, authenticationData);
         }
-
-        Either<Error, Verifier> res = verifierFromBiscuit(role);
-        if (res.isLeft()) {
-            return CompletableFuture.failedFuture(new Exception(res.getLeft().toString()));
-        }
-
-        Verifier verifier = res.get();
-        verifier.set_time();
-        verifier.add_fact(namespaceFact(namespaceName)).get();
-        verifier.add_operation("functions");
-        // should we have #namespace here? why not have #topic for topic rights to be coherent?
-        verifier.add_check("check if right(#authority, #namespace," + namespaceFactFragment(namespaceName) + ",#functions)").get();
-        verifier.allow();
-
-        Either verifierResult = verifier.verify(runLimits);
-        log.debug(verifier.print_world());
-
-        if (verifierResult.isLeft()) {
-            log.debug("Biscuit allowFunctionOpsAsync on [{}] NOT authorized for role [{}]: {}", namespaceName, role, verifierResult.getLeft());
-            return isSuperUser(role, authenticationData, this.conf);
-        } else {
-            log.debug("Biscuit allowFunctionOpsAsync on [{}] authorized.", namespaceName);
-            return CompletableFuture.completedFuture(true);
-        }
+        return isSuperUser(role, authenticationData, this.conf);
```

Comment:
this is only for superuser now?

Reply:
It has been used yet. Will be introduced properly in a next version.